web3/contract.py

```diff
@@ -1328,6 +1337,54 @@ def factory(cls, class_name, **kwargs):
         return PropertyCheckingFactory(class_name, (cls,), kwargs)
 
 
+class ContractCaller:
+    def __init__(self, abi, web3, address, *args, transaction_dict=None, **kwargs):
+        if abi:
+            self.web3 = web3
```

Comment:
I guess `ContractCaller` doesn't have much to do if ABI is `None`. Still, it seems pretty surprising to skip assigning `web3` and `address`, any time `abi` is missing.